### PR TITLE
Several random fixes

### PR DESCRIPTION
**Pull request by mikhailramalho**
This is a big PR that includes several fixed I had scattered in several branches.

The biggest one is the rewrite of polymorphic gcc functions handling: now we create model on the fly during adjust for atomic/sync calls (note that I didn't implement all the models only the ones we "supported" before, I left it as an exercise to someone else). This allowed me to remove the old sync_fetch_add implementation (which only handled int), and the weird atomic_load/atomic_store macros.

Other small changes include:
* Renaming of a few intrinsics that were missing the __ESBMC_ prefix
* Fix printing "error: error: " in cprover_library
* Fix warning in pthread_lib.c and the clang c++ frontend
* Added support for imaxabs (from Rafael)
* Added some C++ template support for name generation
* Removed a bunch of compilers defines that we were setting.
* cleanup our predefined includes
* Fix warning messages to include "WARNING: "
* Make --quiet actually mean quiet, i.e., esbmc prints no output whatsoever
* Moved support of builtins to symex, including the implementation of all the builtin arith checks that we used to support.
* Fixed the "not unwinding loop" in symex that didn't say which loop it wasn't unwinding anymore (this one annoyed me for eons).
* Support overflow exprt in cex

### CODE REVIEW

File: src/goto-symex/execution_state.cpp [hunk 1]
```diff
@@ -242,6 +228,21 @@ void execution_statet::symex_step(reachability_treet &art)
     log_result("{}", oss.str());
   }
 
+  if(break_insn != 0 && break_insn == instruction.location_number)
+  {
+#ifndef _WIN32
+#if !(defined(__arm__) || defined(__aarch64__))
+    __asm__("int $3");
```
Comment (rafaelsamenezes) [inline src/goto-symex/execution_state.cpp]: Would you mind moving this to another module? Something like `asm::breakpoint();`

**Review by rafaelsamenezes — APPROVED** [review]

File: src/goto-symex/execution_state.cpp (hunk 1 above)
Comment (mikhailramalho) [reply src/goto-symex/execution_state.cpp]: sure

**Comment by lucasccordeiro** [comment]
Looks good to me. Just for our sanity, once @mikhailramalho fixes the CI, we should run this PR over the SV-COMP benchmarks.

**Review by lucasccordeiro — APPROVED** [review]
Looks good to me. Just for our sanity, once @mikhailramalho fixes the CI, we should run this PR over the SV-COMP benchmarks.

**Comment by mikhailramalho** [comment]
@rafaelsamenezes, can you test it on sv-comp?

**Comment by rafaelsamenezes** [comment]
@mikhailramalho Started at: https://github.com/esbmc/esbmc/actions/runs/3731512649

With 30s timeout, it should take ~1 day to finish.

**Comment by lucasccordeiro** [comment]
Many thanks, Rafael.Sent from my iPhoneOn 19 Dec 2022, at 08:45, Rafael Sá Menezes ***@***.***> wrote:﻿
@mikhailramalho Started at: https://github.com/esbmc/esbmc/actions/runs/3731512649
With 30s timeout, it should take ~1 day to finish.

**Comment by mikhailramalho** [comment]
> @mikhailramalho Started at: https://github.com/esbmc/esbmc/actions/runs/3731512649
> 
> With 30s timeout, it should take ~1 day to finish.

Did this fail?

**Comment by rafaelsamenezes** [comment]
> Did this fail?

Yeah, it failed while uploading the output folder.

You can check the full log at [here](https://pipelines.actions.githubusercontent.com/serviceHosts/ee7e0430-8afd-4450-8ab8-3f98cafd0ea6/_apis/pipelines/1/runs/17472/signedlogcontent/18?urlExpires=2022-12-21T17%3A45%3A50.1549349Z&urlSigningMethod=HMACV1&urlSignature=pjRlwt5x5Eaa2tAemdYnBaj%2BH0O7l6Ds9ydl9EphsVE%3D)

**Comment by mikhailramalho** [comment]
The log is not enough, as it doesnt show which ones are wrong results.

**Comment by rafaelsamenezes** [comment]
The `[baseline](https://pipelines.actions.githubusercontent.com/serviceHosts/ee7e0430-8afd-4450-8ab8-3f98cafd0ea6/_apis/pipelines/1/runs/17480/signedlogcontent/18?urlExpires=2022-12-26T14%3A19%3A05.6877415Z&urlSigningMethod=HMACV1&urlSignature=8kxtCwn%2F2moY8HaCqOMY36KxzX%B6AsPXaGvXEmBb5sg%3D) has the same negative score.

We need to update the wrapper and add the sv-comp fixes I guess

**Comment by lucasccordeiro** [comment]
@rafaelsamenezes: could you please create such a pull request to update the master branch with the latest SV-COMP fixes?

**Comment by rafaelsamenezes** [comment]
Even after updating the wrapper, it still shows some strange results.

This PR:

```
2022-12-30T23:30:27.4438673Z Statistics:          26501 Files
2022-12-30T23:30:27.4445930Z   correct:           13545
2022-12-30T23:30:27.4446486Z     correct true:     8182
2022-12-30T23:30:27.4447545Z     correct false:    5363
2022-12-30T23:30:27.4462657Z   incorrect:           841
2022-12-30T23:30:27.4463482Z     incorrect true:    771
2022-12-30T23:30:27.4464014Z     incorrect false:    70
2022-12-30T23:30:27.4464358Z   unknown:           12115
2022-12-30T23:30:27.4467717Z   Score:             -4065 (max: 43098)
```

- Logs:  https://github.com/esbmc/esbmc/suites/10100903950/artifacts/493646248

Master:

```
2022-12-30T01:56:22.4778941Z Statistics:          26501 Files
2022-12-30T01:56:22.4784195Z   correct:           13826
2022-12-30T01:56:22.4784789Z     correct true:     8275
2022-12-30T01:56:22.4785618Z     correct false:    5551
2022-12-30T01:56:22.4786589Z   incorrect:           907
2022-12-30T01:56:22.4787150Z     incorrect true:    761
2022-12-30T01:56:22.4791322Z     incorrect false:   146
2022-12-30T01:56:22.4792837Z   unknown:           11768
2022-12-30T01:56:22.4794948Z   Score:             -4587 (max: 43098)
```

- Logs: https://github.com/esbmc/esbmc/suites/10100889631/artifacts/492892921

File: src/esbmc/esbmc_parseoptions.cpp [hunk 2]
```diff
@@ -505,7 +505,8 @@ int esbmc_parseoptionst::doit_k_induction_parallel()
 #ifdef _WIN32
   log_error("Windows does not support parallel kind");
   abort();
-#else
+#endif
```
Comment (fbrausse) [inline src/esbmc/esbmc_parseoptions.cpp:514]: I'd be surprised if this still compiles on Windows.

File: src/goto-symex/symex_goto.cpp [hunk 3]
```diff
@@ -442,18 +442,15 @@ bool goto_symext::get_unwind(
...
-  }
+  const std::string state = stop_unwind ? "Not unwinding " : "Unwinding ";
+
+  log_status(
+    state + "loop {} {} {} {} {}",
+    i2string(cur_state->source.pc->loop_number),
+    " iteration ",
+    integer2string(unwind),
+    " ",
+    cur_state->source.pc->location.as_string());
```
Comment (fbrausse) [inline src/goto-symex/symex_goto.cpp:453]: What a strange format usage with string literals as parameters. Why not this?
```c++
  log_status(
    "{} loop {} iteration {} {}",
    state,
    i2string(cur_state->source.pc->loop_number),
    integer2string(unwind),
    cur_state->source.pc->location.as_string());
```

**Review by fbrausse — COMMENTED** [review]
Impressive cleanup. Though the comment about compilation on Windows probably needs some attention.

**Comment by mikhailramalho** [comment]
fixes already merged as part of other PRs

**Pull request closed without merging**